midterm-2023-11-07/code/5130-stefan-vitanov/script.ts

```diff
@@ -0,0 +1,349 @@
+        }
+ 
+        return 0; 
+      } else if (typeof valA === 'string' && typeof valB === 'string') {
+ 
+        return ascending ? valA.localeCompare(valB) : valB.localeCompare(valA);
+      } else if (typeof valA === 'number' && typeof valB === 'number') {
+ 
+        return ascending ? valA - valB : valB - valA;
+      }
+ 
+ 
+      return 0;
+    });
+  }
+ 
+  async function populateGenreDropdown() {
+    const movies = await getMovies();
+    const genres = new Set(movies.flatMap(movie => movie.genre));
+    const genreSelect = document.getElementById('filter-genre');
+    genres.forEach(genre => {
+      const option = document.createElement('option');
+      option.value = genre;
+      option.textContent = genre;
+      if (genreSelect) {
+        genreSelect.appendChild(option);
+      }
+    });
+  }
+ 
+ 
+  async function populateOscarsDropdown() {
+    const movies = await getMovies();
+    const oscars = new Set(movies.flatMap(movie => Object.keys(movie.oscars)));
+    const oscarSelect = document.getElementById('filter-oscars');
+    oscars.forEach(oscar => {
+      const option = document.createElement('option');
+      option.value = oscar;
+      option.textContent = oscar.replace(/([A-Z])/g, ' $1').trim(); 
+      if (oscarSelect) {
+        oscarSelect.appendChild(option);
+      }
+    });
+  }
+ 
+ 
+  function applyFilters(movies: Movie[]) {
+    const titleElement = document.getElementById('filter-title') as HTMLInputElement;
+    const titleFilter = titleElement ? titleElement.value.toLowerCase() : '';
+ 
+    const yearElement = document.getElementById('filter-year') as HTMLInputElement;
+    const yearFilter = yearElement ? parseInt(yearElement.value, 10) : NaN;
+ 
+    const genreElement = document.getElementById('filter-genre') as HTMLInputElement;
+    const genreFilter = genreElement ? genreElement.value : '';
+ 
+    const oscarElement = document.getElementById('filter-oscars') as HTMLInputElement;
+    const oscarFilter = oscarElement ? oscarElement.value : '';
+ 
+    return movies.filter((movie: Movie) => {
+      const titleMatch = movie.title.toLowerCase().includes(titleFilter);
+      const yearMatch = isNaN(yearFilter) || movie.year === yearFilter;
+      const genreMatch = !genreFilter || movie.genre.includes(genreFilter);
+      const oscarMatch = !oscarFilter || movie.oscars.hasOwnProperty(oscarFilter);
+ 
+      return titleMatch && yearMatch && genreMatch && oscarMatch;
+    });
+  }
+ 
+ 
+  function handleSearchButtonClick() {
+    displayMoviesWithFilters();
+  }
+ 
+  function setUpSearchButton() {
+    const searchButton = document.getElementById('search-button');
+    if (searchButton) {
+      searchButton.addEventListener('click', handleSearchButtonClick);
+    }
+  }
+ 
+ 
+  async function displayMoviesWithFilters() {
+    let movies = await getMovies();
+ 
+ 
+    const titleElement = document.getElementById('filter-title') as HTMLInputElement;
```

Comment:
Now, there should be a way to avoid writing the same code twice :) A call to `applyFilters` will work.